File: chapter4/exercise_33/go/print_processes.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
)

func PrintProcesses() error {

	files, err := os.ReadDir("/proc")
	if err != nil {
		return err
	}

	for _, file := range files {
		if !file.IsDir() {
			continue
		}

		pid := file.Name()

		if _, err := strconv.Atoi(pid); err != nil {
			continue
		}

		commPath := filepath.Join("/proc", pid, "/comm")
		file, err := os.Open(commPath)
		if err != nil {
			continue
		}

		scanner := bufio.NewScanner(file)
		if scanner.Scan() {
			fmt.Printf("PID=%s | Name=%s\n", pid, scanner.Text())
		}

		file.Close()
	}

	return nil
}

func main() {
	PrintProcesses()
}

File: chapter4/exercise_33/cpp/print_processes.cpp
#include <algorithm>
#include <cctype>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <ranges>

void printProcesses()
{
    for (auto &file : std::filesystem::directory_iterator("/proc"))
    {
        const std::string pid = file.path().filename().string();

        if (!std::ranges::all_of(pid, ::isdigit))
        {
            continue;
        }
        std::ifstream commFile(file.path() / "comm");
        std::string processName;

        if (commFile.is_open())
        {
            std::getline(commFile, processName);
            std::cout << "PID: " << pid << " | Name:" << processName << "\n";
        }
        commFile.close();
    }
}

int main()
{

    printProcesses();
    return 0;
}